// @reown/appkit/dist/esm/exports/vue-core.js
import { onUnmounted, ref } from 'vue';
import { ChainController } from '@reown/appkit-controllers';
import { AppKit } from '../src/client/appkit-core.js';
import { getAppKit } from '../src/library/vue/index.js';
import { PACKAGE_VERSION } from './constants.js';
// -- Hooks ------------------------------------------------------------
export * from '../src/library/vue/index.js';
// -- Utils & Other -----------------------------------------------------
export * from '../src/utils/index.js';
export { CoreHelperUtil } from '@reown/appkit-controllers';
let modal = undefined;
export function createAppKit(options) {
    if (!modal) {
        modal = new AppKit({
            ...options,
            basic: true,
            sdkVersion: `vue-core-${PACKAGE_VERSION}`
        });
        getAppKit(modal);
    }
    return modal;
}
export { AppKit };
// -- Hooks ------------------------------------------------------------
export function useAppKitNetwork() {
    const networkState = ChainController.state.activeChain
        ? ChainController.state.chains.get(ChainController.state.activeChain)?.networkState
        : undefined;
    const state = ref({
        caipNetwork: ChainController.state.activeCaipNetwork,
        chainId: ChainController.state.activeCaipNetwork?.id,
        caipNetworkId: ChainController.state.activeCaipNetwork?.caipNetworkId,
        approvedCaipNetworkIds: networkState?.approvedCaipNetworkIds,
        supportsAllNetworks: networkState?.supportsAllNetworks ?? true,
        switchNetwork: async (network) => {
            await modal?.switchNetwork(network);
        }
    });
    const unsubscribes = [];
    unsubscribes.push(ChainController.subscribeKey('activeCaipNetwork', val => {
        state.value.caipNetwork = val;
        state.value.chainId = val?.id;
        state.value.caipNetworkId = val?.caipNetworkId;
        const ns = ChainController.state.activeChain
            ? ChainController.state.chains.get(ChainController.state.activeChain)?.networkState
            : undefined;
        state.value.approvedCaipNetworkIds = ns?.approvedCaipNetworkIds;
        state.value.supportsAllNetworks = ns?.supportsAllNetworks ?? true;
    }));
    unsubscribes.push(ChainController.subscribeChainProp('networkState', networkState => {
        state.value.approvedCaipNetworkIds = networkState?.approvedCaipNetworkIds;
        state.value.supportsAllNetworks = networkState?.supportsAllNetworks ?? true;
    }));
    onUnmounted(() => {
        unsubscribes.forEach(u => u());
    });
    return state;
}
export * from '../src/library/vue/index.js';
