// @coinbase/cdp-sdk/_cjs/_vendor/auth/BearerAuthProvider.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.BearerAuthProvider = void 0;
const core = __importStar(require("../core/index.js"));
const errors = __importStar(require("../errors/index.js"));
const TOKEN_PARAM = "token";
class BearerAuthProvider {
    options;
    constructor(options) {
        this.options = options;
    }
    static canCreate(options) {
        return options?.[TOKEN_PARAM] != null;
    }
    async getAuthRequest({ endpointMetadata, } = {}) {
        const token = await core.EndpointSupplier.get(this.options[TOKEN_PARAM], { endpointMetadata });
        if (token == null) {
            throw new errors.CoinbaseApiError({
                message: BearerAuthProvider.AUTH_CONFIG_ERROR_MESSAGE,
            });
        }
        return {
            headers: { Authorization: `Bearer ${token}` },
        };
    }
}
exports.BearerAuthProvider = BearerAuthProvider;
(function (BearerAuthProvider) {
    BearerAuthProvider.AUTH_SCHEME = "apiKeyAuth";
    BearerAuthProvider.AUTH_CONFIG_ERROR_MESSAGE = `Please provide '${TOKEN_PARAM}' when initializing the client`;
    function createInstance(options) {
        return new BearerAuthProvider(options);
    }
    BearerAuthProvider.createInstance = createInstance;
})(BearerAuthProvider || (exports.BearerAuthProvider = BearerAuthProvider = {}));
