// @coinbase/cdp-sdk/src/_vendor/api/types/DepositTravelRuleOriginator.ts
// This file was auto-generated by Fern from our API Definition.

import type * as CoinbaseApi from "../index.js";

/**
 * Originator information for a deposit travel rule submission.
 */
export interface DepositTravelRuleOriginator {
    /** Full name of the originator. */
    name?: string | undefined;
    address?: CoinbaseApi.PhysicalAddress | undefined;
    /** The type of the originator's wallet. */
    walletType?: DepositTravelRuleOriginator.WalletType | undefined;
    virtualAssetServiceProvider?: CoinbaseApi.DepositTravelRuleVasp | undefined;
    /** Personal identifier for travel rule compliance. For individuals: passport number, national ID, or driver's license. For institutions: LEI (Legal Entity Identifier). */
    personalId?: string | undefined;
    /** Date of birth of the originator. */
    dateOfBirth?: CoinbaseApi.DateOfBirth | undefined;
}

export namespace DepositTravelRuleOriginator {
    /** The type of the originator's wallet. */
    export const WalletType = {
        Custodial: "custodial",
        SelfCustody: "self_custody",
    } as const;
    export type WalletType = (typeof WalletType)[keyof typeof WalletType];
}
