// @coinbase/cdp-sdk/src/_vendor/api/types/Error_.ts
// This file was auto-generated by Fern from our API Definition.

import type * as CoinbaseApi from "../index.js";

/**
 * An error response including the code for the type of error and a human-readable message describing the error.
 */
export interface Error_ {
    errorType: CoinbaseApi.ErrorType;
    /** The error message. */
    errorMessage: string;
    /** A unique identifier for the request that generated the error. This can be used to help debug issues with the API. */
    correlationId?: string | undefined;
    /** A link to the corresponding error documentation. */
    errorLink?: CoinbaseApi.Url | undefined;
    /**
     * The capability code(s) that were not authorized for the customer on
     * this request. Present only when `errorType` is
     * `customer_not_authorized`; absent for every other error type.
     *
     * Use this list to render onboarding UX for the listed capabilities, or
     * fetch `GET /v2/customers/{customerId}` and inspect each entry's
     * `status` / `requirements` to discover what (if anything) can be
     * submitted to resolve the block.
     */
    unauthorizedCapabilities?: CoinbaseApi.CapabilityName[] | undefined;
}
