// @coinbase/cdp-sdk/src/_vendor/api/types/TransferDetails.ts
// This file was auto-generated by Fern from our API Definition.

import type * as CoinbaseApi from "../index.js";

/**
 * Additional details about the transfer. For example, if the transfer was sent to a deposit destination, the information about that destination will be included in this field.
 */
export interface TransferDetails {
    depositDestination?: CoinbaseApi.DepositDestinationReference | undefined;
    /** The onchain transactions associated with the transfer. */
    onchainTransactions?: TransferDetails.OnchainTransactions.Item[] | undefined;
    /** Travel rule compliance status for deposit transfers. Present when the transfer requires travel rule information. */
    travelRule?: TransferDetails.TravelRule | undefined;
}

export namespace TransferDetails {
    export type OnchainTransactions = OnchainTransactions.Item[];

    export namespace OnchainTransactions {
        /**
         * An onchain transaction associated with the transfer.
         */
        export interface Item {
            /** The transaction hash. */
            transactionHash: string;
            network: CoinbaseApi.Network;
        }
    }

    /**
     * Travel rule compliance status for deposit transfers. Present when the transfer requires travel rule information.
     */
    export interface TravelRule {
        status?: CoinbaseApi.TravelRuleStatus | undefined;
        /** Additional details about the current travel rule status. For example, when status is `incomplete`, this may indicate the specific missing information required to proceed. */
        statusMessage?: string | undefined;
    }
}
