// @coinbase/cdp-sdk/src/_vendor/api/types/Transfer.ts
// This file was auto-generated by Fern from our API Definition.

import type * as CoinbaseApi from "../index.js";

/**
 * A Transfer represents all the information needed to execute a transfer and tracks the lifecycle of a transfer from initiation through completion or failure.
 */
export interface Transfer {
    /** The ID of the transfer. Required when validateOnly is false. */
    transferId?: string | undefined;
    status?: CoinbaseApi.TransferStatus | undefined;
    source: CoinbaseApi.TransferSource;
    target: CoinbaseApi.TransferTarget;
    /** The amount of the source asset that will be transferred out, as a decimal string in standard unit denomination. */
    sourceAmount?: string | undefined;
    /** The asset symbol of the source amount. */
    sourceAsset?: CoinbaseApi.Asset | undefined;
    /** The amount of the target asset received, as a decimal string in standard unit denomination. May be omitted in the `quoted` state if the value cannot be guaranteed; see `estimate.targetAmount` for the expected value. Populated with the actual executed amount once the transfer completes. */
    targetAmount?: string | undefined;
    /** The asset symbol of the target amount. */
    targetAsset?: CoinbaseApi.Asset | undefined;
    /** Exchange rate information for currency conversion. The rate indicates how much of the target asset is equivalent to one unit of the source asset. May be omitted in the `quoted` state if the rate cannot be guaranteed; see `estimate.exchangeRate` for the expected rate. Populated with the actual executed rate once the transfer completes. */
    exchangeRate?: CoinbaseApi.TransferExchangeRate | undefined;
    /** The fees associated with this transfer. Different transfer types have different fee structures. May be omitted in the `quoted` state if the fees cannot be guaranteed; see `estimate.fees` for the expected fees. Populated with the actual fees once the transfer completes. */
    fees?: CoinbaseApi.TransferFees | undefined;
    estimate?: CoinbaseApi.TransferEstimate | undefined;
    /** The date and time the transfer was completed. */
    completedAt?: string | undefined;
    /** The reason for failure, if the transfer failed. Only present when status is `failed`. */
    failureReason?: string | undefined;
    /** The date and time when this transfer will expire if not executed. Only present for `quoted` status. A new transfer must be created to obtain an updated quote after expiration. Required when validateOnly is false. */
    expiresAt?: string | undefined;
    /** The date and time the transfer was executed and moved to processing. Only present when status has progressed beyond `quoted`. */
    executedAt?: string | undefined;
    /** The date and time the transfer was created. Required when validateOnly is false. */
    createdAt?: string | undefined;
    /** The date and time the transfer was last updated. Required when validateOnly is false. */
    updatedAt?: string | undefined;
    metadata?: CoinbaseApi.Metadata | undefined;
    details?: CoinbaseApi.TransferDetails | undefined;
}
