// @coinbase/cdp-sdk/src/_vendor/api/types/TransferEstimate.ts
// This file was auto-generated by Fern from our API Definition.

import type * as CoinbaseApi from "../index.js";

/**
 * Captures estimated values for transfers where amounts can't be guaranteed (e.g., USDC -> EURC).
 *
 * The values in `estimate` are not modified after a transfer is executed. They are preserved as an immutable record of the original pre-execution snapshot.
 *
 * The actual executed values are populated in the `transfer` resource post-execution.
 */
export interface TransferEstimate {
    /** The estimated exchange rate at the time this estimate was captured. */
    exchangeRate?: CoinbaseApi.TransferExchangeRate | undefined;
    /** Estimated amount of the target asset that will be received, as a decimal string in standard unit denomination. */
    targetAmount?: string | undefined;
    /** The asset symbol of the estimated target amount. */
    targetAsset?: CoinbaseApi.Asset | undefined;
    /** The estimated fees at the time this estimate was captured. */
    fees?: CoinbaseApi.TransferFees | undefined;
    /** The date and time when this estimate was captured. */
    estimatedAt: string;
}
