// @coinbase/cdp-sdk/_esm/_vendor/api/types/ErrorType.js
// This file was auto-generated by Fern from our API Definition.
/** The code that indicates the type of error that occurred. These error codes can be used to determine how to handle the error. */
export const ErrorType = {
    AlreadyExists: "already_exists",
    AuthorizationExpired: "authorization_expired",
    BadGateway: "bad_gateway",
    CaptureExpired: "capture_expired",
    ClientClosedRequest: "client_closed_request",
    CustomerNotAuthorized: "customer_not_authorized",
    EndpointUnavailable: "endpoint_unavailable",
    FaucetLimitExceeded: "faucet_limit_exceeded",
    Forbidden: "forbidden",
    IdempotencyError: "idempotency_error",
    InternalServerError: "internal_server_error",
    InvalidRequest: "invalid_request",
    InvalidSqlQuery: "invalid_sql_query",
    InvalidSignature: "invalid_signature",
    MalformedTransaction: "malformed_transaction",
    NotFound: "not_found",
    PaymentMethodRequired: "payment_method_required",
    PaymentRequired: "payment_required",
    SettlementFailed: "settlement_failed",
    RateLimitExceeded: "rate_limit_exceeded",
    RequestCanceled: "request_canceled",
    ServiceUnavailable: "service_unavailable",
    TimedOut: "timed_out",
    Unauthorized: "unauthorized",
    UnsupportedTosLanguage: "unsupported_tos_language",
    PolicyViolation: "policy_violation",
    PolicyInUse: "policy_in_use",
    AccountLimitExceeded: "account_limit_exceeded",
    NetworkNotTradable: "network_not_tradable",
    GuestPermissionDenied: "guest_permission_denied",
    GuestRegionForbidden: "guest_region_forbidden",
    GuestTransactionLimit: "guest_transaction_limit",
    GuestTransactionCount: "guest_transaction_count",
    PhoneNumberVerificationExpired: "phone_number_verification_expired",
    OtpVerificationCodeInvalid: "otp_verification_code_invalid",
    OtpVerificationDestinationMismatch: "otp_verification_destination_mismatch",
    OtpVerificationExpired: "otp_verification_expired",
    OtpVerificationInvalid: "otp_verification_invalid",
    OtpVerificationNotFound: "otp_verification_not_found",
    OtpVerificationRequired: "otp_verification_required",
    DocumentVerificationFailed: "document_verification_failed",
    RecipientAllowlistViolation: "recipient_allowlist_violation",
    RecipientAllowlistPending: "recipient_allowlist_pending",
    RefundExpired: "refund_expired",
    TravelRulesRecipientViolation: "travel_rules_recipient_violation",
    SourceAccountInvalid: "source_account_invalid",
    TargetAccountInvalid: "target_account_invalid",
    SourceAccountNotFound: "source_account_not_found",
    TargetAccountNotFound: "target_account_not_found",
    SourceAssetNotSupported: "source_asset_not_supported",
    TargetAssetNotSupported: "target_asset_not_supported",
    TargetEmailInvalid: "target_email_invalid",
    TargetOnchainAddressInvalid: "target_onchain_address_invalid",
    TransferAmountInvalid: "transfer_amount_invalid",
    TransferAssetNotSupported: "transfer_asset_not_supported",
    TransferQuoteExpired: "transfer_quote_expired",
    InsufficientBalance: "insufficient_balance",
    MetadataTooManyEntries: "metadata_too_many_entries",
    MetadataKeyTooLong: "metadata_key_too_long",
    MetadataValueTooLong: "metadata_value_too_long",
    TravelRulesFieldMissing: "travel_rules_field_missing",
    AssetMismatch: "asset_mismatch",
    MfaAlreadyEnrolled: "mfa_already_enrolled",
    MfaInvalidCode: "mfa_invalid_code",
    MfaFlowExpired: "mfa_flow_expired",
    MfaRequired: "mfa_required",
    MfaNotEnrolled: "mfa_not_enrolled",
    OrderQuoteExpired: "order_quote_expired",
    OrderAlreadyFilled: "order_already_filled",
    OrderAlreadyCanceled: "order_already_canceled",
    AccountNotReady: "account_not_ready",
    InsufficientLiquidity: "insufficient_liquidity",
    InsufficientAllowance: "insufficient_allowance",
    TransactionSimulationFailed: "transaction_simulation_failed",
    DelegationNotFound: "delegation_not_found",
    DelegationExpired: "delegation_expired",
    DelegationRevoked: "delegation_revoked",
    DelegationNotAuthorized: "delegation_not_authorized",
    DelegationNotEnabled: "delegation_not_enabled",
    NetworkMismatch: "network_mismatch",
    AlreadyEnabled: "already_enabled",
    PaymentSessionAlreadyCanceled: "payment_session_already_canceled",
    PaymentSessionAlreadyAuthorized: "payment_session_already_authorized",
    PaymentSessionActionPending: "payment_session_action_pending",
    NoCapturableBalance: "no_capturable_balance",
    NoVoidableBalance: "no_voidable_balance",
    NoRefundableBalance: "no_refundable_balance",
    EntityNotConfiguredForPaymentAcceptance: "entity_not_configured_for_payment_acceptance",
    DailyTransactionLimitExceeded: "daily_transaction_limit_exceeded",
    DailyAmountLimitExceeded: "daily_amount_limit_exceeded",
    StaleAttestation: "stale_attestation",
    ModerationRejected: "moderation_rejected",
};
