// @coinbase/cdp-sdk/src/_vendor/api/resources/transfers/client/requests/ListTransfersRequest.ts
// This file was auto-generated by Fern from our API Definition.

import type * as CoinbaseApi from "../../../../index.js";

/**
 * @example
 * ```ts
 * {
 *     accountId: "account_af2937b0-9846-4fe7-bfe9-ccc22d935114",
 *     sourceAccountId: "account_af2937b0-9846-4fe7-bfe9-ccc22d935114",
 *     targetAccountId: "account_af2937b0-9846-4fe7-bfe9-ccc22d935114",
 *     createdAfter: "2026-01-01T00:00:00Z",
 *     createdBefore: "2026-01-31T23:59:59Z",
 *     updatedAfter: "2026-01-01T00:00:00Z",
 *     updatedBefore: "2026-01-31T23:59:59Z",
 *     sourceAsset: "usd",
 *     targetAsset: "usdc",
 *     sourceAddress: "0x742d35Cc6634C0532925a3b844Bc454e4438f44e",
 *     targetAddress: "0x742d35Cc6634C0532925a3b844Bc454e4438f44e",
 *     targetEmail: "user@example.com",
 *     transferId: "transfer_af2937b0-9846-4fe7-bfe9-ccc22d935114",
 *     pageToken: "eyJsYXN0X2lkIjogImFiYzEyMyIsICJ0aW1lc3RhbXAiOiAxNzA3ODIzNzAxfQ=="
 * }
 * ```
 */
export interface ListTransfersRequest {
    /** Filter transfers by status. Useful for building dashboards, monitoring active transfers, or finding transfers needing action. */
    status?: CoinbaseApi.TransferStatus;
    /** Filter transfers by account ID. Returns transfers where the specified account is either the source or target (OR semantics). Cannot be combined with `sourceAccountId` or `targetAccountId`. */
    accountId?: CoinbaseApi.AccountId;
    /** Filter transfers by source account ID. Returns only transfers where the specified account is the source. Cannot be combined with `accountId`. */
    sourceAccountId?: CoinbaseApi.AccountId;
    /** Filter transfers by target account ID. Returns only transfers where the specified account is the target. Cannot be combined with `accountId`. */
    targetAccountId?: CoinbaseApi.AccountId;
    /** Filter transfers to those created at or after this datetime (inclusive). ISO 8601 format. */
    createdAfter?: string;
    /** Filter transfers to those created at or before this datetime (inclusive). ISO 8601 format. */
    createdBefore?: string;
    /** Filter transfers to those updated at or after this datetime (inclusive). ISO 8601 format. Useful for incremental sync — poll for transfers that changed state since your last check. */
    updatedAfter?: string;
    /** Filter transfers to those updated at or before this datetime (inclusive). ISO 8601 format. */
    updatedBefore?: string;
    /** Filter transfers by source asset symbol (e.g., `usd`, `usdc`). */
    sourceAsset?: string;
    /** Filter transfers by target asset symbol (e.g., `usdc`, `eth`). */
    targetAsset?: string;
    /** Filter transfers by the on-chain address of the source. */
    sourceAddress?: CoinbaseApi.BlockchainAddress;
    /** Filter transfers by the on-chain destination address of the target. */
    targetAddress?: CoinbaseApi.BlockchainAddress;
    /** Filter transfers by the email address of the target recipient. */
    targetEmail?: CoinbaseApi.Email;
    /** Filter to a specific transfer by ID. When provided, returns only the matching transfer and bypasses pagination. */
    transferId?: string;
    /** The number of resources to return per page. */
    pageSize?: number;
    /** The token for the next page of resources, if any. */
    pageToken?: string;
}
